cpp/src/arrow/ipc/reader.cc

```diff
@@ -498,19 +498,26 @@ class RecordBatchStreamReader::RecordBatchStreamReaderImpl {
     for (int i = 0; i < dictionary_memo_.num_fields(); ++i) {
       RETURN_NOT_OK(message_reader_->ReadNextMessage(&message));
       if (!message) {
-        /// ARROW-6006: If we fail to find any dictionaries in the stream, then
-        /// it may be that the stream has a schema but no actual data. In such
-        /// case we communicate that we were unable to find the dictionaries
-        /// (but there was no failure otherwise), so the caller can decide what
-        /// to do
-        empty_stream_ = true;
-        break;
+        if (i == 0) {
+          /// ARROW-6006: If we fail to find any dictionaries in the stream, then
+          /// it may be that the stream has a schema but no actual data. In such
+          /// case we communicate that we were unable to find the dictionaries
+          /// (but there was no failure otherwise), so the caller can decide what
+          /// to do
+          empty_stream_ = true;
+          break;
+        } else {
+          // ARROW-6126, the stream terminated before receiving the expected
```

Comment:
I made this point on the java side, but we should clarify the spec.  As I read the spec, as long as a record batch doesn't reference a dictionary they can be interleaved.

Right now this might prematurely return an error if the first record batch has an all null dictionary column, and the following message contains a dictionary for that column.

Reply:
I see. This obviously wouldn't be handled at all in most implementations. One possibility is that dictionaries could be set as length-0 until they are observed (so long as they are all null). Quite esoteric. Agree that ML discussion would be best

Reply:
Sounds good, this isn't a blocker.  Please add the comment on the test method then I think this can be merged.